### PR TITLE
Adding support for defining the .mlgitignore file

### PR DESCRIPTION
**Pull request by lucasncabral**
Inside an entity's directory it is possible that the user has files that he does not want to version. With that in mind, this PR adds support for defining the `.mlgitignore` file which should define rules to indicate which files should be ignored.

The `.mlgitignore` file must be defined in the entity root folder. This means that each entity can have its `.mlgitignore` file, and the defined rules will only be valid for files inside the entity dir.

workspace structure example:
```
project/
├── datasets
│      └── dataset-ex
│          └── data
│              └── file.png
│              └── file.txt
│              └── file2.txt
│              └── augmented
│                  └── file.png
│          └── dataset-ex.spec
│          └── .mlgitignore <- HERE
├── .ml-git
│    ├── datasets
└──  └── config
```

The `.mlgitignore` rules provides support for Unix shell-style wildcards, this rules are similar to the standard `.gitignore` rules. Example of `.mlgitignore` for the worksapce structure above:

```
# Ignore all PNG files inside data directory
data/*.png 

# Ignore specific file
data/file2.txt

# Ignore all JPG files in the entitiy
*.jpg

# Ignore all files inside a folder
data/augmented/
```

The rules defined inside `.mlgitignore`, besides determining which files should not be versioned, also cause these files not to be presented in the status command.

### CODE REVIEW

File: ml_git/file_system/index.py [hunk 1]
```diff
@@ -78,19 +81,21 @@ def _adding_dir_work(self, files, args):
...
-            basepath = root[:len(dirpath) + 1:]
-            relativepath = root[len(dirpath) + 1:]
             for file in files:
-                all_files.append(os.path.join(relativepath, file))
+                file_path = os.path.join(relative_path, file)
+                if ignore_rules is None or not should_ignore_file(ignore_rules, file_path):
+                    all_files.append(file_path)
             self.wp.progress_bar_total_inc(len(all_files))
-            args = {'wp': self.wp, 'basepath': basepath, 'f_index_file': f_index_file, 'all_files': all_files, 'dirpath': dirpath}
+            args = {'wp': self.wp, 'basepath': base_path, 'f_index_file': f_index_file, 'all_files': all_files, 'dirpath': dir_path}
```
Comment (andersonpablosilva) [inline ml_git/file_system/index.py:98]: Shouldn't we rename these keys too (`'basepath'`, `'dirpath'`)?

File: tests/unit/test_utils.py [hunk 2]
```diff
@@ -306,3 +307,24 @@ def test_create_or_update_gitignore(self):
                 for line in file:
                     formatted_line = line.rstrip('\n')
                     self.assertTrue(formatted_line in ignored_files)
+
+    def test_should_ignore_file(self):
+        ignore_rules = ['data/*.png', 'ignored-folder/', '*.jpg', 'data/file-ignored.txt']
```
Comment (andersonpablosilva) [inline tests/unit/test_utils.py:312]: Maybe we can test more complex rules like `data/**/*.ext*`.

**Review by andersonpablosilva — CHANGES_REQUESTED** [review]

File: ml_git/file_system/index.py [hunk 3]
```diff
@@ -78,19 +81,21 @@ def _adding_dir_work(self, files, args):
             return False
         return True
 
-    def _add_dir(self, dirpath, manifestpath, file_path='', trust_links=True):
-        self.manifestfiles = yaml_load(manifestpath)
+    def _add_dir(self, dir_path, manifest_path, file_path='', ignore_rules=None):
```
[the remark this answers is not in the document]
Comment (lucasncabral) [reply ml_git/file_system/index.py]: The method has been moved to be defined before using it. This change was applied in 6a961f1.

File: ml_git/utils.py [hunk 4]
```diff
@@ -389,3 +390,24 @@ def create_or_update_gitignore():
         file.write('\n')
         for remaining_file in ignored_files:
             file.write(remaining_file + '\n')
+
+
+def should_ignore_file(ignore_rules, file_path):
+    if not ignore_rules:
+        return False
+    for ignore in ignore_rules:
+        if fnmatch.fnmatch(file_path, ignore):
```
[the remark this answers is not in the document]
Comment (lucasncabral) [reply ml_git/utils.py]: `ignore` has been replaced by `rule` in 6a961f1. Thanks for the suggestions.

File: ml_git/file_system/index.py (hunk 1 above)
Comment (lucasncabral) [reply ml_git/file_system/index.py]: The keys have been renamed to follow the pattern in 6a961f1.

File: tests/unit/test_utils.py (hunk 2 above)
Comment (lucasncabral) [reply tests/unit/test_utils.py]: Added some new rules for testing: `'data/**/*.tx*'`, `'**/*.test*'`, `'**/access.[0-2].log'`, `'access.[a-c]*'`.

It is important to note that we are using the [fnmatch](https://docs.python.org/3/library/fnmatch.html) lib to perform this comparison between the rules and the file path. The lib is expected to have the proper tests to match the rules.

**Review by andersonpablosilva — APPROVED** [review]

**Review by gilMars — APPROVED** [review]

**Review by Mun1z — APPROVED** [review]

**Review by gpretto — APPROVED** [review]

**Pull request merged**